dotnet/src/Connectors/Connectors.Memory.Postgres/PostgresServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,223 @@
+            {
+                NpgsqlDataSourceBuilder dataSourceBuilder = new(connectionString);
+                dataSourceBuilder.UseVector();
+                return dataSourceBuilder.Build();
+            });
+
+        services.AddKeyedSingleton<IPostgresVectorStoreDbClient>(
+            serviceId,
+            (sp, obj) =>
+            {
+                var dataSource = sp.GetRequiredService<NpgsqlDataSource>();
+                return new PostgresVectorStoreDbClient(dataSource, schema);
+            });
+
+        return services;
+    }
+
+    /// <summary>
+    /// Register a Postgres <see cref="IVectorStore"/> with the specified service ID and where <see cref="IPostgresVectorStoreDbClient"/> is retrieved from the dependency injection container.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="IVectorStore"/> on.</param>
+    /// <param name="options">Optional options to further configure the <see cref="IVectorStore"/>.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <returns>The service collection.</returns>
+    public static IServiceCollection AddPostgresVectorStore(this IServiceCollection services, PostgresVectorStoreOptions? options = default, string? serviceId = default)
+    {
+        // Since we are not constructing the client, add the IVectorStore as transient, since we
+        // cannot make assumptions about how client is being managed.
+        services.AddKeyedTransient<IVectorStore>(
+            serviceId,
+            (sp, obj) =>
+            {
+                var client = sp.GetRequiredService<IPostgresVectorStoreDbClient>();
+                var selectedOptions = options ?? sp.GetService<PostgresVectorStoreOptions>();
+
+                return new PostgresVectorStore(
+                    client,
+                    selectedOptions);
+            });
+
+        return services;
+    }
+
+    /// <summary>
+    /// Register a Postgres <see cref="IVectorStore"/> with the specified service ID and where <see cref="IPostgresVectorStoreDbClient"/> is constructed using the provided parameters.
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/> to register the <see cref="IVectorStore"/> on.</param>
+    /// <param name="connectionString">Postgres database connection string.</param>
+    /// <param name="options">Optional options to further configure the <see cref="IVectorStore"/>.</param>
+    /// <param name="serviceId">An optional service id to use as the service key.</param>
+    /// <returns>The service collection.</returns>
+    public static IServiceCollection AddPostgresVectorStore(this IServiceCollection services, string connectionString, PostgresVectorStoreOptions? options = default, string? serviceId = default)
+    {
+        // Register NpgsqlDataSource to ensure proper disposal.
+        services.AddSingleton<NpgsqlDataSource>(
```

Comment:
Do we need to register this separately?  Is it because it's disposable?

Reply:
That's why I'm registering them separately. Copilot agree'd; copilot output below. But I'm not that familiar with the service collection mechanics so perhaps there's a more straightforward way; if so let me know.

<details>
<summary>Copilot output</summary>

Yes, you need to register the NpgsqlDataSource separately to ensure it is properly disposed of through the service collection. By registering it as a singleton service, you ensure that the ServiceProvider manages its lifecycle and disposes of it when the ServiceProvider itself is disposed.

Explanation
Singleton Registration: Registering NpgsqlDataSource as a singleton ensures that there is only one instance of it throughout the application's lifetime. This is important for managing database connections efficiently.
Proper Disposal: By registering NpgsqlDataSource with the service collection, the ServiceProvider will take care of disposing of it when the application shuts down or when the ServiceProvider is disposed.

</details>